File: Ruby/problem003.rb
# The prime factors of 13195 are 5, 7, 13 and 29.

# What is the largest prime factor of the number 600851475143 ?

def prime_factors(number, array=[])
  2.upto(number) do |factor|
    if number % factor == 0
      array << factor
      number /= factor
      prime_factors(number, array)
      break
    end
  end
  array
end

p prime_factors(600851475143).last


# 100

# 2..100

# 100/2

# [2], 50

# 50

# 2..50

# [2,2], 25

# 25

# [2,2,5], 5

[2,2,5,5], 1

File: CSharp/problem003.cs
//The prime factors of 13195 are 5, 7, 13 and 29.

//What is the largest prime factor of the number 600851475143 ?

using System;
using System.Collections.Generic;
using System.Linq;

namespace Problem003
{
   class MainClass
   {
      public static void Main()
      {
         List<long> factorArray = new List<long>();
         Console.WriteLine(PrimeFactors(600851475143, factorArray));
      }

      public static long PrimeFactors(long number, List<long> factorArray)
      {
         for (long i = 2; i <= number; i++)
         {
            if (number % i == 0) {
               factorArray.Add(i);
               number = number / i;
               PrimeFactors(number, factorArray);
               break;
            }
         }
         factorArray.Sort();
         return factorArray.Last();
      }
   }
}
